draft-ietf-tls-esni.md

```diff
@@ -476,6 +473,37 @@ contain a "cached_info" extension {{!RFC7924}} with a CachedObject
 entry whose CachedInformationType is "cert", since this indication
 would divulge the true server name.
 
+## Recommended Padding Scheme {#padding}
+
+Future extensions could reveal sensitive information through their length. Consequently, 
+padding should be flexible and support arbitrary extension changes. 
+
+For each field in the inner ClientHello, clients need to determine how much to
+pad given the semantics of that field.  For example, if a client can propose
+various ALPN values, it could add padding to round up to the longest of those.
+
+The target padding length of most ClientHello extensions can be determined without server help.
+However, the "server_name" extension could benefit from server input (ECHOConfig.maximum_name_length). 
+Clients SHOULD compute the padding for this extension as follows:
+
+1. If ECHOConfig.maximum_name_length is longer than the actual server_name then 
+clients SHOULD add padding to make up that difference.
+2. Otherwise, if ECHOConfig.maximum_name_length is zero or less than the length 
+of the actual server_name then round the server_name up to a multiple of 32 octets 
+and randomly add another 32 octets 50% of the time and then include that amount
+of additional padding.
+
+The amount of padding applied to the ClientHello is then computed as the sum of 
+all per-extension padding values, rounded up to the nearest multiple of 32. 
+(This additional rounding step aims to hide variance across different client
+implementations.)
+
+In addition to padding ClientHelloInner, clients and servers will also need
+to pad all other handshake messages that have sensitive-length fields. For
+example, if a client proposes ALPN values in ClientHelloInner, the
+server-selected value will be returned in an EncryptedExtension, so that
+handshake message also needs to be padded using TLS record layer padding.
```

Comment:
Is there a good reference for "TLS record layer padding"? I think I know what this means, but it could be clearer.

Reply:
It's defined in Section 5.4 of RFC8446: https://tools.ietf.org/html/rfc8446#section-5.4

Reply:
Maybe that reference suggests a terminology issue. It says "All encrypted TLS records can be padded to inflate the size of the TLSCiphertext." Does ECHO define things in terms of TLSCiphertext?

I don't necessarily think there is a problem here, but in the two code bases I've read (rustls and NSS), the encrypted SNI or ClientHello haven't fit in all that cleanly.